server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+or the duration may be inherently variable. In such cases the client may want to
+block for _up to_ a client-chosen duration after which time the issuing client
+disconnects from the RPC but the update processing continues on. The system must
+support the client re-attaching to the Update at some point in the future in
+order to observe the UpdateResponse. Note that this is different than putting a
+timeout on the RPC itself.
+
+### Known Long Update
+
+Similar to the uncertain duration, it may be know a priori that the update will
+take more time than the issuer is willing to block the thread of execution
+invoking the update. In such cases the update issuer would prefer to not wait at
+all for the UpdateResponse, preferring instead to poll for it later.
+
+### Invalid Update
+
+UpdateRequests can be invalid given the state of the receiving workflow. In such
+cases the UpdateResponse will contain an error indication and payload. Invalid
+updates _must not_ be recorded in workflow history to prevent a malicious issuer
+from causing a denial of service on the workflow by issuing repeated invalid
+UpdateRequests.
+
+### "Failing" Update
+
+UpdateRequests can be semantically acceptable but still fail due to external
+dependencies. In such cases the UpdateResponse will contain an error indication
+and payload.
+
+### Unexpected Disconnect
+
+While the issuer of an update may wish to block on an UpdateResponse, network
+conditions or software failures may occur which forcibly disconnects the update
+issuer from the server at the time that the Update completes _and_ before the
+issuer becomes aware of the Update ID. The issuer must be able to re-issue the
+same UpdateRequest to access the server-generated Update ID.
+
+## gRPC API
+
+The gRPC API comprises two functions, `UpdateWorkflow` and
+`AwaitUpdateWorkflowResponse` to be added to the `WorkflowService` service
+definition. The former is used to invoke an update and follows the standard
+patterns for identifying a workflow execution and carrying arbitrary metadata
+(Header) and payloads (Payloads). The request type for this function includes a
+`wait_timeout` property indicating how long the client is willing to wait for a
+synchronous response once the request has been accepted by the system. This
+value can be set to zero indicating no such waiting should occur.  The
+`request_id` property uniquely identifies the logical request (_not_ the
+issuance of the request) and is intended for use as an idempotency token.
+
+The second function, `AwaitUpdateWorkflowResponse` is used when a client has
+already issued an `UpdateWorkflow` request and has the resulting `update_id`.
+That `update_id` can be provided along with a `wait_timeout` (same semantics as
+above) to re-attach to an in-flight or completed Update and obtain the
+associated UpdateResult in the form of an `UpdateWorkflowResponse`.
+
+
+```protobuf
+service WorkflowService {
```

Comment:
We have talked about having separate endpoints for long poll vs. short lived requests.
It makes more sense for latency metrics on these endpoints.

We should consider adding another StartUpdateWorkflow RPC that schedules the update and returns the update_id ASAP.

Reply:
This is a good point that I had neglected to deal with. I went back through the use cases and ended up in a slightly different place. In the latest version, UpdateWorkflow will always be (in a rough sense) fast. The client either wants the response "inline" with the gRPC response, in which case we clamp the timeout to a couple of seconds at max, or the client only waits for update_id and will poll later on. I think/hope this lets us get away with just the one UpdateWorkflow call.

As for metrics, given that UpdateWorkflow will invoke workflow code in all cases, my inclination is to exclude this function from latency metrics as it is out of Temporal's control.